File: shaper/src/main/java/com/osama/shaper/FragmentComponent.java
package com.osama.shaper;

import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;


public abstract class FragmentComponent<T extends Fragment> {
    private
    ComponentView componentView;
    private boolean isContentViewSet;


    private T fragment;
    private LayoutInflater inflater;

    public void onCreateView(@NonNull T fragment, LayoutInflater inflater, @Nullable Bundle mSavedInstanceState) {
        this.fragment = fragment;
        this.inflater = inflater;
    }

    public void onViewCreated(@NonNull T fragment, View view, @Nullable Bundle mSavedInstanceState) {

    }


    public void onResume(@NonNull T fragment) {

    }

    public void onStop(T fragment) {

    }

    public ComponentView getComponentView() {
        return componentView;
    }

    @Nullable
    protected View setContentView(@LayoutRes int layout) {
        if (componentView != null && !isContentViewSet) {
            isContentViewSet = true;
            return LayoutInflater.from(fragment.getActivity()).inflate(layout, componentView);
        } else return null;
    }

    public FragmentComponent<T> setComponentView(ComponentView componentView) {
        if (this.componentView == null) this.componentView = componentView;
        return this;
    }
}

File: shaper/src/main/java/com/osama/shaper/FragmentComponent.kt
package com.osama.shaper

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.support.annotation.IdRes
import android.support.annotation.LayoutRes
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import io.reactivex.disposables.CompositeDisposable


abstract class FragmentComponent<T : Fragment> {

    private var isContentViewSet: Boolean = false


    protected lateinit var fragment: T

    private var inflater: LayoutInflater? = null

    protected val context: Context?
        get() = fragment.context

    protected val activity: Activity?
        get() = fragment.activity

    protected val disposables = CompositeDisposable()

    open fun onCreate(mSavedInstanceState: Bundle?) {}

    open fun onCreateView(fragment: T, inflater: LayoutInflater, mSavedInstanceState: Bundle?) {
        this.fragment = fragment
        this.inflater = inflater
    }

    open fun onViewCreated(fragment: T, view: View, mSavedInstanceState: Bundle?) {}

    open fun onResume(fragment: T) {

    }

    open fun onStop(fragment: T) {
        disposables.clear()
        disposables.dispose()
    }

    open fun onStart(fragment: T) {

    }

    open fun onDestroy(fragment: T) {

    }

    protected fun setContentView(@LayoutRes layout: Int, @IdRes hostLayout: Int): View? {
        return if (!isContentViewSet) {
            isContentViewSet = true
            LayoutInflater.from(context).inflate(layout, activity?.findViewById(hostLayout))
        } else null
    }
}
